Build rdiff target optionally

mbrt opened the pull request:
I've tweaked the build process a bit. I have the need to minimize build dependencies, and so I made the following changes:
* Added an option (default ON) to build the `rdiff` target;
* Added an option (default OFF) to build with compression support;
* Made `zlib` and `bzip2` libraries optional.

This allows me to build only the `rsync` library (and not `rdiff`), and thus avoid having `popt` dependency. Moreover, `zlib` and `bzip2` use can be disabled.

The removal of `REQUIRED` attribute in `find_package` turns an hard error in a soft one, and does not stop the generation step. This allows the user to tweak parameters after generation with CMake.

The compression support is not implemented for now, so I've kept it OFF by default. When the support is present, you can make the option ON by default!

Let me know what you think.

mbrt commented:
Mmm... Not sure about the failure under clang-apple. I don't think it's my fault. Master seems broken too.

sourcefrog commented:
Sounds good to me.

> Mmm... Not sure about the failure under clang-apple in travis. I don't think it's my fault. Master seems broken too.

Yes, that's #50, not your fault.

sourcefrog commented on CMakeLists.txt:
```diff
@@ -37,6 +37,14 @@ if (NOT CMAKE_SYSTEM_NAME)
+# Add an option to exclude rdiff executable from build
+# This is useful, because it allows to remove POPT dependency if a user is interested only in the
+# rsync library itself and not in the rdiff executable
+option(BUILD_RDIFF "Whether or not to build rdiff executable" ON)
```
Please document this in `install.md`

sourcefrog commented on CMakeLists.txt:
```diff
@@ -37,6 +37,14 @@ if (NOT CMAKE_SYSTEM_NAME)
+# rsync library itself and not in the rdiff executable
+option(BUILD_RDIFF "Whether or not to build rdiff executable" ON)
+
+# Add an option to include compression support
```
Please mention https://github.com/librsync/librsync/issues/8 here, that it's not actually functional

sourcefrog commented on CMakeLists.txt:
```diff
@@ -37,6 +37,14 @@ if (NOT CMAKE_SYSTEM_NAME)
+# This is useful, because it allows to remove POPT dependency if a user is interested only in the
+# rsync library itself and not in the rdiff executable
+option(BUILD_RDIFF "Whether or not to build rdiff executable" ON)
+
```
Please try adding in .travis.yml a config that builds without `rdiff`.

sourcefrog commented on CMakeLists.txt:
```diff
@@ -262,15 +287,20 @@ install(TARGETS rsync ${INSTALL_TARGETS_DEFAULT_ARGS} DESTINATION lib)
+endif (BUILD_RDIFF)
 
 
 ########### install files ###############
```
Please mention this and credit yourself in NEWS.md

sourcefrog commented:
That looks like a really worthwhile change, thanks. I'd just like a few tweaks.

mbrt commented:
I've addressed your comments. There is only one additional change. I've disabled some tests when `BUILD_RDIFF` is disabled, because they depend on `rdiff` executable! This is unfortunate, but the alternative would be to have those tests to fail... Let me know!

sourcefrog commented:
Yeah, unfortunately past-me wrote most of the tests on top of rdiff rather than as C-level unit tests. Probably better to disable them for now though it does make this a bit more risky. Maybe we should note that in install.md?

mbrt commented:
I've disabled them in e215215. I think the risk is not very much, since the default build is with `rdiff`, and we have also travis which tests all the configurations. However it's better to add a note in install.md, as you said!

mbrt commented:
Added the note in the last commit.

sourcefrog commented:
Great, thanks.
